db/schema.rb

```diff
@@ -145,6 +145,8 @@
     t.integer "role_id"
     t.datetime "last_login"
     t.boolean "secure_password", default: false, null: false
+    t.integer "failed_attempts"
+    t.datetime "last_failed_attempt"
```

Comment:
Doesn't this cause the failed_attempts counter to just increase, if for example there is one failed attempt every 23 hours, because `last_failed_attempt` gets reset to current time again and again, while `failed_attempts` is not being reset?

Imho this implementation has bunch of corner cases where it will not behave as documented/desired, and needs a rewrite.

Reply:
That's why the counter is reset here: https://github.com/bigbluebutton/greenlight/pull/3019/files#diff-9802ca3c9c4cf89904fd44bc114e35ebdf2c5dd3d5b645491e2b253e1afef29bR229

> Imho this implementation has bunch of corner cases where it will not behave as documented/desired, and needs a rewrite.

If you can provide me with those edge cases, I'd be glad to look over the implementation :)

Reply:
The counter is only reset if the last failed attempt is more than 24 hours ago. So, a Denial of Service attack would be to for example generate a list of students email addresses, then lock all of their accounts using 5 invalid login requests each, then once every 23 hours send one invalid request again

Reply:
Alternatively, send one failed login per 23 hours and after 5 days it will resut in account being locked and continue being locked as long as a request keeps coming once every 23 hours

Reply:
@basisbit is absolutely right and it's even what the comment says:

```
# reset the counter if the last login attempt was more than 1 day ago
update(failed_attempts: 0) if attempts.positive? && !within_1_day
``` 

Edge case: User does not remember his password and tries it several times until he gets block. A couple of hours he tries it again and the timeout starts from the beginning. At the next morning he starts again, which restarts the time once again.

In my opinion the waiting time should be related to the number of failed attempts to a maximum. E.g. the waiting time in minutes could be something like `last_failed_attempt + max(60, failed_attempts * 5)`.

Reply:
@sualko fyi, I already reported that as DoS vulnerability to Blindside Networks. I hope it gets fixed soon-ish.